### DISCUSSION TITLE
Are checkpoints and controlnets cached between different API calls?

### DISCUSSION THREAD

**Discussion by allo-** [comfyanonymous/ComfyUI#4096] [Q&A]
When running a workflow, particularly via the API, are the checkpoints and controlnets unloaded between two calls? If not, how many different checkpoints stay loaded and when are they unloaded? Does Comfy unload everything directly or is there some kind of garbage collection?

I'm looking for the most efficient way to automate comfyui workflows, and the node concept looks like it might start from zero each time, at least for the API calls. Also, I am not sure how checkpoint caching is handled in general, as they are not loaded in advance before running the workflow the first time.

**Comment by ltdrdata** [comment]
It is not unloaded between API calls. Any nodes that do not exist with the same id and type between the execution of the first prompt and the second prompt are treated as deleted nodes.

And if there are deleted nodes or nodes whose widget values have changed, the cache of all nodes affected by the results of those nodes is invalidated.

If you want to maintain the checkpoint and controlnet models between two prompt executions, there are two approaches:

1. Keep those two nodes with completely identical IDs.
2. Use something like the backend cache nodes from the Inspire Pack to store and use them in a separate cache.

**Comment by allo-** [comment]
Are the IDs in the API JSON meant to be unique between different API calls? When I export a workflow they are usually small integers, so there would be a large risk of collisions when I have two saved workflows from different Comfy runs.

**Reply by ltdrdata** [reply]
The assignment of IDs is done by the side that requests the API. 
The backend only uses the combination of ID and node type to determine whether a node has changed between two prompt execution requests.

If you send the same prompt twice without making any modifications, the backend will determine that there are no changes and will do nothing.

**Comment by allo-** [comment]
My use case is that I have a few workflows with different prompts and Controlnet inputs that run frequently. If I understand using LoRA in Comfy correctly, changing LoRA will also need to load a new model to the GPU. Not sure if loading the model from CPU RAM to VRAM is the limiting factor though. Maybe having the model in disk cache is already quite good.

I'll see what works and have a look at the Inspire nodes. Thanks for your help so far.

**Reply by ltdrdata** [reply]
FYI, https://www.youtube.com/watch?v=7wfe1DBAkoA

https://github.com/ltdrdata/ComfyUI-extension-tutorials/blob/Main/ComfyUI-Inspire-Pack/workflow/cache-workflow1.png
https://github.com/ltdrdata/ComfyUI-extension-tutorials/blob/Main/ComfyUI-Inspire-Pack/workflow/cache-workflow2.png

The original purpose of creating these nodes was to prevent model reloading while switching workflows.
